playground/backend/internal/code_processing/code_processing.go

```diff
@@ -268,56 +334,75 @@ func finishByTimeout(ctx context.Context, pipelineId uuid.UUID, cacheService cac
 	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_RUN_TIMEOUT)
 }
 
-// processError processes error received during processing code via setting a corresponding status and output to cache
-func processError(ctx context.Context, err error, data []byte, pipelineId uuid.UUID, cacheService cache.Cache, status pb.Status) {
-	switch status {
-	case pb.Status_STATUS_VALIDATION_ERROR:
-		logger.Errorf("%s: Validate: %s\n", pipelineId, err.Error())
+// processValidateError processes error received during processing validate step via setting a corresponding status and output to cache
+func processValidateError(ctx context.Context, errorChannel chan error, pipelineId uuid.UUID, cacheService cache.Cache) {
+	err := <-errorChannel
+	logger.Errorf("%s: Validate: %s\n", pipelineId, err.Error())
+
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_VALIDATION_ERROR)
+}
+
+// processPrepareError processes error received during processing prepare step via setting a corresponding status and output to cache
+func processPrepareError(ctx context.Context, errorChannel chan error, pipelineId uuid.UUID, cacheService cache.Cache) {
+	err := <-errorChannel
+	logger.Errorf("%s: Prepare: %s\n", pipelineId, err.Error())
+
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_PREPARATION_ERROR)
+}
 
-		cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_VALIDATION_ERROR)
-	case pb.Status_STATUS_PREPARATION_ERROR:
-		logger.Errorf("%s: Prepare: %s\n", pipelineId, err.Error())
+// processCompileError processes error received during processing compile step via setting a corresponding status and output to cache
+func processCompileError(ctx context.Context, errorChannel chan error, errorOutput []byte, pipelineId uuid.UUID, cacheService cache.Cache) {
+	err := <-errorChannel
+	logger.Errorf("%s: Compile: err: %s, output: %s\n", pipelineId, err.Error(), errorOutput)
 
-		cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_PREPARATION_ERROR)
-	case pb.Status_STATUS_COMPILE_ERROR:
-		logger.Errorf("%s: Compile: err: %s, output: %s\n", pipelineId, err.Error(), data)
+	cacheService.SetValue(ctx, pipelineId, cache.CompileOutput, "error: "+err.Error()+", output: "+string(errorOutput))
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_COMPILE_ERROR)
+}
```

Comment:
processCompileError and processRunError also very similar, try to think how to merge them (maybe even with processPrepareError and processValidateError)

Reply:
`processRunError` contains logic with stopping goroutine before setting a new status to the cache. It works using 2 channels. I don't think that it is a good idea to merge `processRunError` and `processCompileError` methods because `processCompileError` doesn't have this logic and it doesn't use these channels.